### ISSUE TITLE
Packmap should be able to find SystemJS-compatible dist files

### ISSUE THREAD

**Issue by brandones** [openmrs/packmap#12]
The `main` file specified by `package.json`s is often a CommonJS module. This is incompatible with SystemJS. Packmap should be able to identify UMD (or ES? or AMD?) files and put those in the import-map instead.

**Comment by joeldenning** [comment]
### Question

Which module(s) did you try to load that were CJS? I'm guessing 3p libs?

### Related context

See related https://github.com/openmrs/packmap/pull/1#discussion_r330208740 where @blittle suggested using the package.json `module` field for this.

The bundler / module loader world in javascript has had many many many debates about how to handle this. Webpack and rollup both respect the package.json `module` field for this, where `main` means "maybe CJS maybe ESM" and `module` means "definitely ESM". However, there is also the [package.json "browser"](https://docs.npmjs.com/files/package.json#browser) field, which is actually sorta just ignored by most bundlers and module loaders. See [this good discussion in webpack](https://github.com/webpack/webpack/issues/4674) for more info on how those are being used. Generally speaking, whatever webpack supports is what all the third party libs start implementing. If it's not supported by webpack, not as many packages actually will modify their package.json to do it.

In addition to the frontend browser bundler discussions, NodeJS settled on how it would know the difference between CJS and ESM. It eventually arrived at the [package.json "type" field](https://nodejs.org/api/esm.html#esm_code_package_json_code_code_type_code_field) to indicate if a package is an ES module or a CJS module, along with [other ways of enabling ESM](https://nodejs.org/api/esm.html#esm_enabling).

I'm sharing all this ^^ to give you an idea of what we're dealing with here when we make a decision for packmap. The jspm project is somewhat similar to packmap and it has its own set of rules about how it process packages, as decided by @guybedford's deep knowledge on the topic. I created packmap because jspm and tracemap (two existing projects) couldn't quite do what was needed for OpenMRS. However, we should be aware of the standards and patterns evolving in the frontend community so that packmap doesn't go down a weird path.

### The confounding variable with packmap

Since packmap is only used by OpenMRS, it is 100% geared towards modules that are consumable by SystemJS, as configured in [master-single-page-application.jsp](https://github.com/openmrs/openmrs-module-spa/blob/master/omod/src/main/webapp/master-single-page-application.jsp#L14-L19). This means that the following module formats are the ones usable by the OpenMRS community:

| Module type | Supported by OpenMRS |
| ------------- | ---------------------------------------  |
| [System.register](https://github.com/systemjs/systemjs/blob/master/docs/system-register.md) | ✅ |
| [AMD / UMD](https://github.com/amdjs/amdjs-api/wiki/AMD) | ✅ |
| Global variable | ✅ |
| JSON | ✅ |
| [CommonJS](http://www.commonjs.org/) | ❌ |
| [ESM](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) | ❌ |

Note that, since ESM is not supported, I do not think using the package.json `"module"` field would make sense here. Since as far as I know, there is no single aspect of a package.json that will indicate to us which file to load that will always be supported by OpenMRS, I created the [package-json-overrides directory](https://github.com/openmrs/packmap/tree/master/lib/package-json-overrides). Note that this is similar to the [jspm registry](https://github.com/jspm/registry/tree/master/package-overrides/npm), except less complete and sophisticated since it only has a few packages and doesn't have support for package versioning.

The easiest way to solve the problem that caused you to open this issue would probably be just to add a package override to the directory I linked to above. Now that I've explained the full context, though, do you agree? Or have a different approach that would work?

**Comment by mseaton** [comment]
@joeldenning and @brandones - I'm new to this, so please point me to any basic resources I might need to better get up to speed if I'm asking overly basic questions.  I spoke briefly today with Brandon today about some issues I have had getting things up and running locally (various JS console errors around exports not defined, etc), and it sounded like this issue might be the source of at least some of my issues, so I figured I'd post here.

If I am understanding what you are saying Joel, you are recommending that for now we need to manually manage these issues in our package-json-overrides file rather than in our packages.json file.  This would be like the following:

`{
  "imports": {
    "@openmrs/esm-root-config": "/openmrs/frontend/@openmrs/esm-root-config@1.1.2/dist/openmrs-esm-root-config.defaults.js",
    "i18next": "/openmrs/frontend/i18next@19.0.0/dist/umd/i18next.js"
  }
}`

Is this correct?

I think if so that could be ok, but we should aim for better error messages to let users know that they are encountering these issues, and better documentation around known problematic libraries that exist (eg. i18next).

It might also be nice to have these dependencies auto-update where appropriate.  Is that possible without causing other problems?  Eg. if I update my package.json to i18next 19.0.1, this would be reflected in the override configuration as specified above.

**Comment by joeldenning** [comment]
> manually manage these issues in our package-json-overrides file rather than in our packages.json file

Yes, that's what I'm thinking and is what is set up in packmap so far.

> we should aim for better error messages to let users know that they are encountering these issues

Packmap is a build tool and it doesn't know at build time whether it will work once it hits the browser. When the error occurs in the browser, it is thrown by SystemJS. Brandon created https://github.com/systemjs/systemjs/issues/2082 to try to improve the error messages there.

> It might also be nice to have these dependencies auto-update where appropriate. Is that possible without causing other problems?

Yeah that is how it is set up, as long as the i18next package itself doesn't change which files it publishes.

**Comment by guybedford** [comment]
I'd suggest looking into the package.json "exports" field pattern here, as documented in the Node.js es modules documentation. A custom `"system"` definition for System modules might be useful in this, saving having to define a new top-level package.json field.
